straxen/plugins/acquisition_monitor.py

```diff
@@ -0,0 +1,109 @@
+    hev tag <= DDC10 hardware high energy veto
+    """
+        
+    __version__ ='0.0.2'
+    depends_on = ('raw_records_aqmon') 
+    provides  = ('veto_intervals',)
+    data_kind = ('veto_intervals',)
+
+    aqmon_channel_names = ('sum_wf','m_veto_sync','hev_stop', 'hev_start', 'bb_stop', 'bb_start','stop', 'start')
+    duration_dtype = [(('Interval between start and stop in [ns]', 'interval'), np.int64)]
+    
+    # Using the time & endtime dtypes for busy so strax does not complain about lack of endtime and time dtypes
+    def infer_dtype(self):
+        dtype = []
+        for veto in ('','hev_','bb_'):
+            dtype += [(('Start '+ veto +'veto time since unix epoch [ns]', veto+'time'), np.int64)]
+            dtype += [(('Stop '+ veto +'veto time since unix epoch [ns]', veto+'endtime'), np.int64)]
+            dtype += [(('Duration of '+ veto +'veto time since unix epoch [ns]', veto+'interval'), np.int64)]
+        return dtype
+    
+    def setup(self):
+        self.channel_range = self.config['channel_map']['aqmon']
+        self.channel_numbers = np.arange(self.channel_range[0]+1,self.channel_range[1]+1,1)
+        self.channel_map = dict(zip(self.aqmon_channel_names,self.channel_numbers))
+        return self.channel_map
+    
+    def get_window_size(self):
+        return (self.config['max_veto_gap'])    
+    
+    
+    def compute(self, raw_records_aqmon): 
+        r = raw_records_aqmon
+        vetos = dict()
+        
+        for i, v in enumerate(['','bb_','hev_']):
+            channels = channel_select(r,self.channel_map[v+'stop'], self.channel_map[v +'start'])
+            vetos[v +'veto'] = merge_vetos(channels,gap = self.config['max_veto_gap'],\
+                                           dtype = strax.time_fields + self.duration_dtype)
+            result = strax.dict_to_rec(vetos)
+        
+        busy_interval = result['veto']
+        bb_interval = result['bb_veto']
+        hev_interval = result['hev_veto']
+        
+        return dict(time = busy_interval['time'],
+                   endtime = busy_interval['endtime'],
+                   interval = busy_interval['interval'],
+                   hev_time = hev_interval['time'],
+                   hev_endtime = hev_interval['endtime'],
+                   hev_interval = hev_interval['interval'],
+                   bb_time = bb_interval['time'],
+                   bb_endtime = bb_interval['endtime'],
+                   bb_interval = bb_interval['interval'])
+    
+    @staticmethod   
+    def merge_vetos(channels, gap, dtype):
+        if len(channels):
+            start, stop = strax.find_peak_groups(channels, gap_threshold = gap)  
+            result = np.zeros(len(start),dtype=dtype)
+            result['time'] = start
+            result['endtime'] = stop
+            result['interval'] = stop - start
+        else:
+            result = np.zeros(1,dtype=dtype)
+            result['time'] = 0
```

Comment:
This is somewhat probamatic in strax. It does all kinds of checks and this way it will have data before the chunck is written. Maybe set it to the first time of the chunck?

Reply:
You mean setting the times to 0?

Reply:
Yeah the field 'time' is special in strax. What you get now is:
```
Exception in thread build:veto_intervals:
Traceback (most recent call last):
  File "/home/xedaq/miniconda/envs/py37/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/home/xedaq/miniconda/envs/py37/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/xedaq/software/strax/strax/mailbox.py", line 262, in _send_from
    self.kill_from_exception(e)
  File "/home/xedaq/software/strax/strax/mailbox.py", line 186, in kill_from_exception
    raise e
  File "/home/xedaq/software/strax/strax/mailbox.py", line 254, in _send_from
    x = next(iterable)
  File "/home/xedaq/software/strax/strax/plugin.py", line 514, in iter
    yield from super().iter(iters, executor=executor)
  File "/home/xedaq/software/strax/strax/plugin.py", line 363, in iter
    yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
  File "/home/xedaq/software/strax/strax/plugin.py", line 531, in do_compute
    result = super().do_compute(chunk_i=chunk_i, **kwargs)
  File "/home/xedaq/software/strax/strax/plugin.py", line 431, in do_compute
    return self._fix_output(result, start, end)
  File "/home/xedaq/software/strax/strax/plugin.py", line 460, in _fix_output
    data=result)
  File "/home/xedaq/software/strax/strax/plugin.py", line 478, in chunk
    data=data)
  File "/home/xedaq/software/strax/strax/chunk.py", line 71, in __init__
    raise ValueError(f"Attempt to create chunk {self} "
ValueError: Attempt to create chunk [007161.veto_intervals: 1584603686sec 999998998 ns - 1584603693sec 999999000 ns, 1 items, 0.0 MB/s] whose data starts early at 0

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-23-3e13f5cb7623> in <module>
      1 # Trying to open the plugin
      2 data_custom = st.get_array(run_id, ['veto_intervals'],
----> 3                           register=[VetoIntervals])

~/software/strax/strax/context.py in get_array(self, run_id, targets, save, max_workers, **kwargs)
    862                 max_workers=max_workers,
    863                 **kwargs)
--> 864             results = [x.data for x in source]
    865         return np.concatenate(results)
    866 

~/software/strax/strax/context.py in <listcomp>(.0)
    862                 max_workers=max_workers,
    863                 **kwargs)
--> 864             results = [x.data for x in source]
    865         return np.concatenate(results)
    866 

~/software/strax/strax/context.py in get_iter(self, run_id, targets, save, max_workers, time_range, seconds_range, time_within, time_selection, selection_str, _chunk_number, **kwargs)
    765                 allow_lazy=self.context_config['allow_lazy'],
    766                 max_messages=self.context_config['max_messages'],
--> 767                 timeout=self.context_config['timeout']).iter()):
    768             seen_a_chunk = True
    769             if not isinstance(result, strax.Chunk):

~/software/strax/strax/chunk.py in continuity_check(chunk_iter)
    249     last_end = None
    250     last_runid = None
--> 251     for s in chunk_iter:
    252         if s.run_id != last_runid:
    253             # TODO: can we do better?

~/software/strax/strax/processor.py in iter(self)
    260             # which is printed for the user
    261             self.log.debug("Reraising exception")
--> 262             raise exc.with_traceback(traceback)
    263 
    264         # Check the savers for any exception that occurred during saving

~/software/strax/strax/mailbox.py in _send_from(self, iterable)
    252 
    253                 try:
--> 254                     x = next(iterable)
    255                 except StopIteration:
    256                     # No need to send this yet, close will do that

~/software/strax/strax/plugin.py in iter(self, iters, executor)
    512 
    513     def iter(self, iters, executor=None):
--> 514         yield from super().iter(iters, executor=executor)
    515 
    516         # Yield final results, kept at bay in fear of a new chunk

~/software/strax/strax/plugin.py in iter(self, iters, executor)
    361                 yield new_future
    362             else:
--> 363                 yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
    364 
    365         raise RuntimeError("This cannot happen.")

~/software/strax/strax/plugin.py in do_compute(self, chunk_i, **kwargs)
    529 
    530         # Compute new results
--> 531         result = super().do_compute(chunk_i=chunk_i, **kwargs)
    532 
    533         # Throw away results we already sent out

~/software/strax/strax/plugin.py in do_compute(self, chunk_i, **kwargs)
    429         result = self.compute(**kwargs)
    430 
--> 431         return self._fix_output(result, start, end)
    432 
    433     def _fix_output(self, result, start, end, _dtype=None):

~/software/strax/strax/plugin.py in _fix_output(self, result, start, end, _dtype)
    458                 end=end,
    459                 data_type=_dtype,
--> 460                 data=result)
    461         return result
    462 

~/software/strax/strax/plugin.py in chunk(self, start, end, data, data_type, run_id)
    476             data_type=data_type,
    477             dtype=self.dtype_for(data_type),
--> 478             data=data)
    479 
    480     def compute(self, **kwargs):

~/software/strax/strax/chunk.py in __init__(self, data_type, data_kind, dtype, run_id, start, end, data)
     69 
     70             if data_starts_at < self.start:
---> 71                 raise ValueError(f"Attempt to create chunk {self} "
     72                                  f"whose data starts early at {data_starts_at}")
     73             if data_ends_at > self.end:

ValueError: Attempt to create chunk [007161.veto_intervals: 1584603686sec 999998998 ns - 1584603693sec 999999000 ns, 1 items, 0.0 MB/s] whose data starts early at 0

```

Reply:
That is strange, because I run it fine in my jupyter notebook on midway. The  = 0 solution was supposed to be a "trick" to avoid dealing with no data to supply, hence its checking the lenght of data first.

Reply:
Interesting, you probably have the latest strax and straxen versions, I guess? Maybe the particular run you are looking at never ends up here and the one I happen to look at does. Nevertheless, we should also handle these cases correctly.

Reply:
Yes, indeed I run the latest strax and straxen versions. I was using run 007158 for all the testing, I was not sure if others even have AM data.

Reply:
I will try with another run.

Reply:
@jorana  With run 00761 I can't event get raw_records_aqmon, it just gives this below. I tried to fiddle with my context to make daqreader work, but I guess will be better to look at it tomorrow morning.

```
Removing data in ./strax_data/007161-raw_records-rfzvpzj4mf to overwrite
Removing data in ./strax_data/007161-raw_records_he-rfzvpzj4mf to overwrite
Removing data in ./strax_data/007161-raw_records_mv-rfzvpzj4mf to overwrite
Removing data in ./strax_data/007161-raw_records_aqmon-rfzvpzj4mf to overwrite
Removing data in ./strax_data/007161-veto_intervals-i5wrpniu3t to overwrite
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
Exception in thread read_0:DAQReader_divide_outputs_mailbox:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/mailbox.py", line 486, in divide_outputs
    result = next(source)
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: DAQReader_divide_outputs_mailbox did not get 0 in time.
Exception in thread build:veto_intervals:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/mailbox.py", line 262, in _send_from
    self.kill_from_exception(e)
  File "/home/elykov/strax/strax/mailbox.py", line 186, in kill_from_exception
    raise e
  File "/home/elykov/strax/strax/mailbox.py", line 254, in _send_from
    x = next(iterable)
  File "/home/elykov/strax/strax/plugin.py", line 514, in iter
    yield from super().iter(iters, executor=executor)
  File "/home/elykov/strax/strax/plugin.py", line 265, in iter
    _fetch_chunk(d)
  File "/home/elykov/strax/strax/plugin.py", line 247, in _fetch_chunk
    [self.input_buffer[d], next(iters[d])])
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: raw_records_aqmon_mailbox did not get 0 in time.


Exception in thread save_0:raw_records_aqmon:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/storage/common.py", line 572, in save_from
    source.throw(e)
  File "/home/elykov/strax/strax/storage/common.py", line 547, in save_from
    chunk = next(source)
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: raw_records_aqmon_mailbox did not get 0 in time.
Exception in thread save_0:veto_intervals:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/storage/common.py", line 572, in save_from
    source.throw(e)
  File "/home/elykov/strax/strax/storage/common.py", line 545, in save_from
    [chunk, next(source)])
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: veto_intervals_mailbox did not get 0 in time.


Exception in thread save_0:raw_records_mv:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/storage/common.py", line 572, in save_from
    source.throw(e)
  File "/home/elykov/strax/strax/storage/common.py", line 547, in save_from
    chunk = next(source)
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: raw_records_mv_mailbox did not get 0 in time.
Exception in thread save_0:raw_records:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/storage/common.py", line 572, in save_from
    source.throw(e)
  File "/home/elykov/strax/strax/storage/common.py", line 547, in save_from
    chunk = next(source)
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: raw_records_mailbox did not get 0 in time.

Exception in thread save_0:raw_records_he:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/storage/common.py", line 572, in save_from
    source.throw(e)
  File "/home/elykov/strax/strax/storage/common.py", line 547, in save_from
    chunk = next(source)
  File "/home/elykov/strax/strax/mailbox.py", line 358, in _read
    f"{self.name} did not get {next_number} in time.")
strax.mailbox.MailboxReadTimeout: raw_records_he_mailbox did not get 0 in time.


DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
DAQReader:140535149704752 waiting for chunk 0
Exception in thread divide_outputs:raw_records_aqmon:
Traceback (most recent call last):
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/elykov/.conda/envs/daq/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/elykov/strax/strax/mailbox.py", line 262, in _send_from
    self.kill_from_exception(e)
  File "/home/elykov/strax/strax/mailbox.py", line 186, in kill_from_exception
    raise e
  File "/home/elykov/strax/strax/mailbox.py", line 254, in _send_from
    x = next(iterable)
  File "/home/elykov/strax/strax/plugin.py", line 282, in iter
    f"{self.__class__.__name__}:{id(self)} waited for "
strax.plugin.InputTimeoutExceeded: DAQReader:140535149704752 waited for more  than 80 sec for arrival of input chunk 0, and has given up.

---------------------------------------------------------------------------
MailboxReadTimeout                        Traceback (most recent call last)
<ipython-input-26-066057475924> in <module>()
      1 st.make('007158', 'veto_intervals')
----> 2 acq_interval = st.get_df('007161', 'veto_intervals')

~/strax/strax/context.py in get_df(self, run_id, targets, save, max_workers, **kwargs)
    873         df = self.get_array(
    874             run_id, targets,
--> 875             save=save, max_workers=max_workers, **kwargs)
    876         try:
    877             return pd.DataFrame.from_records(df)

~/strax/strax/context.py in get_array(self, run_id, targets, save, max_workers, **kwargs)
    862                 max_workers=max_workers,
    863                 **kwargs)
--> 864             results = [x.data for x in source]
    865         return np.concatenate(results)
    866 

~/strax/strax/context.py in <listcomp>(.0)
    862                 max_workers=max_workers,
    863                 **kwargs)
--> 864             results = [x.data for x in source]
    865         return np.concatenate(results)
    866 

~/strax/strax/context.py in get_iter(self, run_id, targets, save, max_workers, time_range, seconds_range, time_within, time_selection, selection_str, _chunk_number, **kwargs)
    765                 allow_lazy=self.context_config['allow_lazy'],
    766                 max_messages=self.context_config['max_messages'],
--> 767                 timeout=self.context_config['timeout']).iter()):
    768             seen_a_chunk = True
    769             if not isinstance(result, strax.Chunk):

~/strax/strax/chunk.py in continuity_check(chunk_iter)
    249     last_end = None
    250     last_runid = None
--> 251     for s in chunk_iter:
    252         if s.run_id != last_runid:
    253             # TODO: can we do better?

~/strax/strax/processor.py in iter(self)
    260             # which is printed for the user
    261             self.log.debug("Reraising exception")
--> 262             raise exc.with_traceback(traceback)
    263 
    264         # Check the savers for any exception that occurred during saving

~/strax/strax/mailbox.py in _send_from(self, iterable)
    252 
    253                 try:
--> 254                     x = next(iterable)
    255                 except StopIteration:
    256                     # No need to send this yet, close will do that

~/strax/strax/plugin.py in iter(self, iters, executor)
    512 
    513     def iter(self, iters, executor=None):
--> 514         yield from super().iter(iters, executor=executor)
    515 
    516         # Yield final results, kept at bay in fear of a new chunk

~/strax/strax/plugin.py in iter(self, iters, executor)
    263         _end = float('inf')
    264         for d in self.depends_on:
--> 265             _fetch_chunk(d)
    266             if self.input_buffer[d].end < _end:
    267                 pacemaker = d

~/strax/strax/plugin.py in _fetch_chunk(d, hope_to_see)
    245                 # print(f"Fetching {d} in {self}, hope to see {hope_to_see}")
    246                 self.input_buffer[d] = strax.Chunk.concatenate(
--> 247                     [self.input_buffer[d], next(iters[d])])
    248                 # print(f"Fetched {d} in {self}, "
    249                 #      f"now have {self.input_buffer[d]}")

~/strax/strax/mailbox.py in _read(self, subscriber_i)
    356                                                          self.timeout):
    357                         raise MailboxReadTimeout(
--> 358                             f"{self.name} did not get {next_number} in time.")
    359                 self._subscriber_waiting_for[subscriber_i] = None
    360 

MailboxReadTimeout: raw_records_aqmon_mailbox did not get 0 in time.
```

Reply:
Hmmm .... I think you are right, the 0s do introduce quite a lot of issues. However, at the moment I don't know how to to do it differently. Any help would be appreciated.